### PR TITLE
1.0.0

### LINKED ISSUE #5: have a special key when injecting the container itself
When you inject the container it self into a dependency there should be a special key which can be used. This would allow the container to inject the correct  container into the dependency. This is really for usage with child container, a dependency resolved from a child container that takes a container should be injected with that child container.

### LINKED ISSUE #13: Allow groups to be injected as an array to resolved objects
If you do this:

```
var groupIdentifier ='myGroupId';
container
  .register('identifier1', FoosCtorFunction)
  .inject('otherDependencyIdentifier1')
  .inGroup(groupIdentifier);
container
  .register('identifier2', BarsCtorFunction)
  .inject('otherDependencyIdentifier1')
  .inGroup(groupIdentifier);
```

you should be able to inject an array of the group `groupIdentifier` items via `inject`:

```
var groupIdentifier ='myGroupId';
container
  .register('identifier3', BazzCtorFunction)
  .inject(groupIdentifier);
```

In this case an array of `[item1:FoosCtorFunction, item2:BarsCtorFunction]` would be passed to a `BazzCtorFunction` instance upon resolution/construction. 

The rules for creation of items in the array should follow that of `container.resolveGroup(groupIdentifier)`

### LINKED ISSUE #14: should throw if you register a null or undefined instance via `container.registerInstance()`

### PR DESCRIPTION
**Pull request by KeithWoods**
Items covered in 1.0.0: 
* Closes #14 - should throw if you register a null or undefined instance via `container.registerInstance()`
* Closes #13 - Allow groups to be injected as an array to resolved objects
* Closes #5 - have a special key when injecting the container itself
* Additional error checking on public api calls. 
* Version bumped to 1.0.0, the container has had a lot of usage now and is stable. 
* ReadMe updated with new features.

### CODE REVIEW

File: microdi-js.d.ts [hunk 1]
```diff
@@ -36,4 +36,7 @@ export class RegistrationModifier {
 
 export interface Resolver<T> {
     resolve(container:Container, dependencyKey:any):T
+}
+export class MicroDiConsts {
+    static get owningContainer();
```
Comment (raybooysen) [inline microdi-js.d.ts]: missing the type?

File: src/Container.js [hunk 2]
```diff
@@ -45,13 +47,15 @@ export default class Container {
         child._resolvers = Object.create(this._resolvers);
         child._isDisposed = false;
         child._childContainers = [];
+        child._registerSelf();
```
Comment (raybooysen) [inline src/Container.js]: Shouldn't this be opt-in?

File: src/Container.js [hunk 3]
```diff
@@ -73,12 +79,19 @@ export default class Container {
...
     isRegistered(name) {
+        Guard.isNonEmptyString(name, 'Error calling isRegistered(name). The name argument must be a string and can not be \'\'');
         this._throwIfDisposed();
         var registration = this._registrations[name];
         return !!registration;
     }
-
+    isGroupRegistered(groupName) {
+        Guard.isNonEmptyString(groupName, 'Error calling isGroupRegistered(groupName). The groupName argument must be a string and can not be \'\'');
+        this._throwIfDisposed();
```
Comment (raybooysen) [inline src/Container.js]: Should it not throw first?  Why bother with the guards if it's disposed?

File: src/Guard.js [hunk 4]
```diff
@@ -13,12 +13,12 @@
...
  notice_end */
- 
+
 import * as utils from './utils';
 
 export default class Guard {
-    static isDefined(value, message) {
-        if (typeof value === 'undefined') {
+    static isNotNullOrUndefined(value, message) {
+        if (typeof value === 'undefined' || value === null) {
```
Comment (raybooysen) [inline src/Guard.js]: `if(value != null)` should do the trick?

https://codepen.io/anon/pen/pRjarN?editors=1111

File: microdi-js.d.ts (hunk 1 above)
Comment (KeithWoods) [inline microdi-js.d.ts]: oops, will add `:string`

File: src/Container.js (hunk 2 above)
Comment (KeithWoods) [inline src/Container.js]: I don't think so. This aids resolution of the correct containers for objects being resolved. There should be no side effects of doing this and all it really does is put a 'built in' register  into the container. I think it's important to be a default.

File: src/Container.js (hunk 3 above)
Comment (KeithWoods) [inline src/Container.js]: yep, good point. happens in a few places and can fix

File: src/Guard.js (hunk 4 above)
Comment (KeithWoods) [inline src/Guard.js]: yeah I had thought of using that exact code but thought of being more explicit

**Comment by raybooysen** [comment]
Looks good to me. :)

**Pull request merged**